### PR TITLE
Refactored cleaning.py, altered test file

### PR DESCRIPTION
**Pull request by joaogomes98**

### CODE REVIEW

File: life_expectancy/cleaning.py [hunk 1]
```diff
@@ -26,6 +20,17 @@ def clean_data(country):
...
+    """
+    Function responsible for cleaning the data and 
+    filtering it according to a specific region.
+
+    The data cleaning follows this logic: the years are converted 
+    into integers and the values are converted into floats
+    """
+
     # Convert year to int
     dataframe['year'] =  pd.to_numeric(dataframe['year'], errors='coerce').astype(int)
```
Comment (nfpaiva) [inline life_expectancy/cleaning.py:36]: suggestion: perhaps you could have set up a function to take care of all data types

image

File: life_expectancy/cleaning.py [hunk 2]
```diff
@@ -56,16 +67,18 @@ def str_to_float(val):
...
+    """
+
+    dataframe = load_data('life_expectancy/data/eu_life_expectancy_raw.tsv')
 
-     # Create argument parser
-    parser = argparse.ArgumentParser(description='Clean life expectancy \
-        data for a specified country')
-    parser.add_argument('--country', type=str, default='PT', help='ISO code \
-        of country to filter by')
+    dataframe = clean_data(dataframe, 'PT')
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:78]: You are right, my mistake. The main should indeed receive the country as an argument and pass it to the clean_data method. Thanks!

File: life_expectancy/cleaning.py [hunk 3]
```diff
@@ -56,16 +67,18 @@ def str_to_float(val):
...
 
-    # Parse arguments
-    args = parser.parse_args()
+    save_data(dataframe, 'life_expectancy/data/pt_life_expectancy.csv')
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:80]: Nice suggestion! It makes total sense.

File: life_expectancy/tests/test_cleaning.py [hunk 4]
```diff
@@ -1,13 +1,13 @@
...
 
-from life_expectancy.cleaning import clean_data
 from . import OUTPUT_DIR
 
 
 def test_clean_data(pt_life_expectancy_expected):
-    """Run the `clean_data` function and compare the output to the expected output"""
-    clean_data('PT')
+    """Run the main function and compare the output to the expected output"""
+    cleaning.main()
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/tests/test_cleaning.py:10]: You are right, my mistake. The main should indeed receive the country as an argument and pass it to the clean_data method. Thanks!

File: life_expectancy/cleaning.py [hunk 5]
```diff
@@ -56,16 +67,18 @@ def str_to_float(val):
         # If the conversion fails, return NaN
         return np.nan
```
[the remark this answers is not in the document]
Comment (MrCordeiro) [reply life_expectancy/cleaning.py:69]: In this case, calling `group()` on `None` would raise an AttributeError. Luckily, João is already catching that error (❤️), but mypy isn't smart enough to notice that (yet).

IMO, this bit looks good as it is, but I understand that getting mypy warnings might be a bother, so there are 2 alternative:

1. To run `matches = re.search(r'\d+\.*\d*', val)`  in the line above and the use a `if matches:` condition
2. To use Pandas' [pandas.Series.str.replace](https://pandas.pydata.org/docs/reference/api/pandas.Series.str.replace.html)

File: life_expectancy/cleaning.py [hunk 6]
```diff
@@ -35,13 +40,19 @@ def clean_data(country):
...
-    # Save cleaned data to csv
-    dataframe.to_csv('life_expectancy/data/pt_life_expectancy.csv', index=False)
+    return dataframe
+
+def save_data(dataframe: pd.DataFrame, filepath: str) -> None:
+    """
+    Function responsible for saving the data to a file.
+    """
+
+    dataframe.to_csv(filepath, index=False)
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:53]: Doesn't hurt to have some error handling ;)

File: life_expectancy/cleaning.py [hunk 7]
```diff
@@ -1,19 +1,13 @@
 import re
-import argparse
 import pandas as pd
 import numpy as np
 
-def clean_data(country):
+def load_data(filepath: str) -> pd.DataFrame:
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:5]: Thats odd... When I submitted my code for review, I made sure I had 10.0 in pylint. I ran it again, with some of the new changes made, and it is giving me a 10 again. I am going to ignore this for now.

File: life_expectancy/cleaning.py [hunk 8]
```diff
@@ -1,19 +1,13 @@
...
-    filtering it according to a specific region.
-
-    The data cleaning follows this logic: the years are converted 
-    into integers and the values are converted into floats
+    Function responsible for loading the data.
     """
 
-    # Load data
-    dataframe = pd.read_csv('life_expectancy/data/eu_life_expectancy_raw.tsv', delimiter='\t')
+    dataframe = pd.read_csv(filepath, delimiter='\t')
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:10]: Good suggestion!

File: life_expectancy/cleaning.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:69]: I made the changes to the code so it uses the str.replace() method. I believe there are no warnings now

File: life_expectancy/cleaning.py [hunk 9]
```diff
@@ -26,6 +20,17 @@ def clean_data(country):
...
+    """
+    Function responsible for cleaning the data and 
+    filtering it according to a specific region.
+
+    The data cleaning follows this logic: the years are converted 
+    into integers and the values are converted into floats
+    """
+
     # Convert year to int
     dataframe['year'] =  pd.to_numeric(dataframe['year'], errors='coerce').astype(int)
```
[the remark this answers is not in the document]
Comment (joaogomes98) [reply life_expectancy/cleaning.py:35]: With the erros='coerce', I believe any sort of error is handled and turned into a NaN. I don't know, from reading the assignment, if that is the expected behavior, but I also noticed that I should improve the 'removing any NaNs' logic to include the year column

File: life_expectancy/cleaning.py (hunk 1 above)
Comment (joaogomes98) [reply life_expectancy/cleaning.py:36]: Nice suggestion. Makes sense to specify the types for each column, just to make sure everything is correct

File: life_expectancy/cleaning.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (nfpaiva) [reply life_expectancy/cleaning.py:6]: ok, but the point is, should we've docstring at beginning of each module?

PEP8 says yes :) https://peps.python.org/pep-0257/

image

File: life_expectancy/cleaning.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (nfpaiva) [reply life_expectancy/cleaning.py]: suggestion: great you did the change however PEP8 recommends uppercase on constants such as paths.

image

**Review by nfpaiva — COMMENTED** [review]
just two minor PEP8 comments, for me that's ok to merge after you decide if you want to push the suggestions.

**Pull request merged**